File: backend/domains/map/services/data/moonPhase.js
// Moon phase calculator for real 1950 lunar data.
// Used by climate API, trip-day generator, and narrative prompts.

// Known new moon for 1950-01-18 ~08:00 UTC (NASA/USNO tables)
const KNOWN_NEW_MOON_UTC = new Date('1950-01-18T08:00:00Z');
const SYNODIC_MONTH_DAYS = 29.530588861;

function toDate(input) {
  if (input instanceof Date) return input;
  if (typeof input === 'string') {
    const match = input.match(/^(\d{4}-\d{2}-\d{2})/);
    if (match) return new Date(`${match[1]}T00:00:00Z`);
  }
  return new Date(input);
}

/**
 * Compute the moon phase for a given date.
 * @param {Date|string} dateInput - a Date or an ISO/date string (YYYY-MM-DD or timestamp)
 * @returns {{phase: string, illumination: number, age_days: number}}
 */
export function getMoonPhase(dateInput) {
  const d = toDate(dateInput);
  const daysSince = (d - KNOWN_NEW_MOON_UTC) / (1000 * 60 * 60 * 24);
  const synodicMonth = SYNODIC_MONTH_DAYS;
  const age = ((daysSince % synodicMonth) + synodicMonth) % synodicMonth;
  const illumination = 0.5 * (1 - Math.cos(2 * Math.PI * age / synodicMonth));

  const newWindow = 1.5;
  const fullWindow = 1.5;
  const quarterWindow = 1.5;
  const q = synodicMonth / 4;

  let phase;
  if (age < newWindow || age > synodicMonth - newWindow) {
    phase = 'new_moon';
  } else if (Math.abs(age - q) < quarterWindow) {
    phase = 'first_quarter';
  } else if (Math.abs(age - 2 * q) < fullWindow) {
    phase = 'full_moon';
  } else if (Math.abs(age - 3 * q) < quarterWindow) {
    phase = 'last_quarter';
  } else if (age < q) {
    phase = 'waxing_crescent';
  } else if (age < 2 * q) {
    phase = 'waxing_gibbous';
  } else if (age < 3 * q) {
    phase = 'waning_gibbous';
  } else {
    phase = 'waning_crescent';
  }

  return {
    phase,
    illumination,
    age_days: age,
  };
}

File: story-engine/app/moon_phase.py
# ============================================================================
# Moon phase calculator for real 1950 lunar data.
# ----------------------------------------------------------------------------
# Port of backend/domains/map/services/data/moonPhase.js (pure function, no DB).
# ============================================================================
import math
import re
from datetime import datetime, timezone

KNOWN_NEW_MOON_UTC = datetime(1950, 1, 18, 8, 0, 0, tzinfo=timezone.utc)
SYNODIC_MONTH_DAYS = 29.530588861


def _to_date(input_):
    if isinstance(input_, datetime):
        return input_ if input_.tzinfo else input_.replace(tzinfo=timezone.utc)
    if isinstance(input_, str):
        match = re.match(r'^(\d{4}-\d{2}-\d{2})', input_)
        if match:
            return datetime.strptime(match.group(1), '%Y-%m-%d').replace(tzinfo=timezone.utc)
    raise ValueError(f'Unsupported date input: {input_!r}')


def get_moon_phase(date_input):
    """Compute the moon phase for a given date.

    Returns { phase, illumination, age_days }.
    """
    d = _to_date(date_input)
    days_since = (d - KNOWN_NEW_MOON_UTC).total_seconds() / (60 * 60 * 24)
    synodic_month = SYNODIC_MONTH_DAYS
    age = ((days_since % synodic_month) + synodic_month) % synodic_month
    illumination = 0.5 * (1 - math.cos(2 * math.pi * age / synodic_month))

    new_window = 1.5
    full_window = 1.5
    quarter_window = 1.5
    q = synodic_month / 4

    if age < new_window or age > synodic_month - new_window:
        phase = 'new_moon'
    elif abs(age - q) < quarter_window:
        phase = 'first_quarter'
    elif abs(age - 2 * q) < full_window:
        phase = 'full_moon'
    elif abs(age - 3 * q) < quarter_window:
        phase = 'last_quarter'
    elif age < q:
        phase = 'waxing_crescent'
    elif age < 2 * q:
        phase = 'waxing_gibbous'
    elif age < 3 * q:
        phase = 'waning_gibbous'
    else:
        phase = 'waning_crescent'

    return {'phase': phase, 'illumination': illumination, 'age_days': age}
